File: Sources/LoafCat/Modules/MessageModule.swift
import AppKit

/// Two ways to make the cat carry a message: one that arrives at a time, and one
/// that stays until it is taken down.
///
/// Both go through the same bubble, so a note and a reminder can never overlap; the
/// reminder wins while it is up and the note comes back underneath it.
final class MessageModule: CatModule {
    let id = "message"

    private let bus: WellnessBus

    /// The minute the reminder last fired in, so a 120Hz tick cannot fire it 7200
    /// times during the minute it is due.
    private var firedOnDay: Int = -1

    private var demoFireAt: CFAbsoluteTime = .greatestFiniteMagnitude
    private var meowStart: CFAbsoluteTime = -1
    private let meowDuration: Double = 1.4
    private let showSeconds: Double = 8

    init(bus: WellnessBus) {
        self.bus = bus
    }

    /// `--demo-timers` cannot wait for a wall-clock minute to roll around.
    func armDemoReminder(after seconds: Double) {
        demoFireAt = CFAbsoluteTimeGetCurrent() + seconds
        if bus.settings.reminderText.isEmpty {
            bus.settings.reminderText = "Stand up and look out of a window."
        }
    }

    func update(_ ctx: TickContext) -> ModuleOutput {
        let now = CFAbsoluteTimeGetCurrent()

        if now >= demoFireAt {
            demoFireAt = .greatestFiniteMagnitude
            fire(bus.settings.reminderText, why: "demo")
        } else if bus.settings.reminderEnabled, !bus.busy {
            checkClock(now: now, secondsSinceKey: ctx.secondsSinceKey)
        }

        guard meowStart >= 0 else { return .none }
        let p = (now - meowStart) / meowDuration
        guard p < 1 else { meowStart = -1; return .none }

        // The meow: a head-forward lunge and a wobble. There is no open-mouth cell
        // in the rig and inventing one would need art in three themes, so the
        // gesture carries it.
        var out = ModuleOutput()
        let t = CGFloat(p)
        let envelope = (1 - t) * (1 - t)
        out.squash = 1 + envelope * 0.07 * sin(.pi * 4 * t)
        out.offset.y = -envelope * bus.atlas.wellness.bobHeight * 0.8 * sin(.pi * 2 * t)
        return out
    }

    private func checkClock(now: CFAbsoluteTime, secondsSinceKey: Double) {
        let parts = bus.settings.reminderTime.split(separator: ":")
        guard parts.count == 2, let hh = Int(parts[0]), let mm = Int(parts[1]) else { return }

        let cal = Calendar.current
        let date = Date()
        let c = cal.dateComponents([.hour, .minute, .dayOfYear, .year], from: date)
        guard let hour = c.hour, let minute = c.minute, let day = c.dayOfYear,
              let year = c.year else { return }
        let dayKey = year * 1000 + day
        guard hour == hh, minute == mm, firedOnDay != dayKey else { return }
        firedOnDay = dayKey

        if secondsSinceKey > bus.awaySeconds {
            // Marked as fired, then dropped: a reminder is about a moment, and
            // replaying it an hour later is worse than not having it.
            bus.log(String(format: "reminder SKIPPED, away %.0fs", secondsSinceKey))
            return
        }
        fire(bus.settings.reminderText, why: "clock \(bus.settings.reminderTime)")
    }

    private func fire(_ text: String, why: String) {
        let message = text.isEmpty ? "Reminder!" : text
        bus.bubble?.say(message, for: showSeconds)
        bus.chime()
        meowStart = CFAbsoluteTimeGetCurrent()
        bus.log("reminder (\(why)) \"\(message)\"")
    }

    // MARK: - pinned note

    func pin(_ text: String?) {
        bus.bubble?.pin(text)
        bus.log("note \(text == nil ? "cleared" : "pinned: \"\(text!)\"")")
    }

    func clearReminder() {
        bus.settings.reminderEnabled = false
        bus.settings.reminderTime = ""
    }

    // MARK: - dialogs

    func promptForReminder() {
        // An .accessory app has no menu bar focus of its own; without this the alert
        // opens behind whatever the user is working in.
        NSApp.activate(ignoringOtherApps: true)

        let alert = NSAlert()
        alert.messageText = "Scheduled reminder"
        alert.informativeText = "The cat will meow and hold this up at the time you set."
        alert.addButton(withTitle: "Save")
        alert.addButton(withTitle: "Cancel")

        let time = NSTextField(string: bus.settings.reminderTime.isEmpty
            ? Self.defaultTimeString() : bus.settings.reminderTime)
        time.placeholderString = "HH:MM (24-hour)"
        let body = NSTextField(string: bus.settings.reminderText)
        body.placeholderString = "Message"

        let stack = NSStackView(views: [labelled("Time", time), labelled("Message", body)])
        stack.orientation = .vertical
        stack.spacing = 8
        stack.frame = NSRect(x: 0, y: 0, width: 280, height: 64)
        alert.accessoryView = stack
        alert.window.initialFirstResponder = time

        guard alert.runModal() == .alertFirstButtonReturn else { return }
        guard Self.parse(time.stringValue) != nil else {
            warn("That time did not parse. Use HH:MM, 24-hour — for example 14:30.")
            return
        }
        bus.settings.reminderTime = Self.normalise(time.stringValue)
        bus.settings.reminderText = body.stringValue
        bus.settings.reminderEnabled = true
        firedOnDay = -1
    }

    func promptForNote() {
        NSApp.activate(ignoringOtherApps: true)
        let alert = NSAlert()
        alert.messageText = "Pin a note"
        alert.informativeText = "It stays above the cat until you unpin it."
        alert.addButton(withTitle: "Pin")
        alert.addButton(withTitle: "Cancel")

        let field = NSTextField(string: bus.bubble?.hasPinnedNote == true
            ? bus.settings.pinnedNote : "")
        field.placeholderString = "Note"
        field.frame = NSRect(x: 0, y: 0, width: 280, height: 24)
        alert.accessoryView = field
        alert.window.initialFirstResponder = field

        guard alert.runModal() == .alertFirstButtonReturn else { return }
        pin(field.stringValue)
    }

    private func labelled(_ title: String, _ field: NSTextField) -> NSView {
        let label = NSTextField(labelWithString: title)
        label.setContentHuggingPriority(.required, for: .horizontal)
        label.alignment = .right
        label.setFrameSize(NSSize(width: 60, height: 20))
        let row = NSStackView(views: [label, field])
        row.orientation = .horizontal
        row.spacing = 8
        return row
    }

    private func warn(_ text: String) {
        let a = NSAlert()
        a.messageText = text
        a.alertStyle = .warning
        a.runModal()
    }

    private static func parse(_ s: String) -> (Int, Int)? {
        let parts = s.trimmingCharacters(in: .whitespaces).split(separator: ":")
        guard parts.count == 2, let h = Int(parts[0]), let m = Int(parts[1]),
              (0...23).contains(h), (0...59).contains(m) else { return nil }
        return (h, m)
    }

    private static func normalise(_ s: String) -> String {
        guard let (h, m) = parse(s) else { return s }
        return String(format: "%02d:%02d", h, m)
    }

    private static func defaultTimeString() -> String {
        let c = Calendar.current.dateComponents([.hour, .minute], from: Date())
        return String(format: "%02d:%02d", c.hour ?? 9, c.minute ?? 0)
    }
}

File: windows/LoafCat/Modules/MessageModule.cs
using System.Globalization;
using System.Runtime.Versioning;

namespace LoafCat.Modules;

/// Two ways to make the cat carry a message: one that arrives at a time, and one that
/// stays until it is taken down.
///
/// Both go through the same bubble, so a note and a reminder can never overlap; the
/// reminder wins while it is up and the note comes back underneath it.
///
/// The macOS file also carries `promptForReminder` and `promptForNote`, two NSAlert
/// dialogs left over from when the menu bar owned these settings. Nothing calls them
/// there any more — Settings does both — so they are not ported. The parser they
/// shared with the settings field is, because that one IS still shared.
[SupportedOSPlatform("windows")]
public sealed class MessageModule(WellnessBus bus) : ICatModule
{
    public string Id => "message";

    private readonly WellnessBus _bus = bus;

    /// The day the reminder last fired on, so a 120Hz tick cannot fire it 7200 times
    /// during the minute it is due.
    private int _firedOnDay = -1;

    private double _demoFireAt = Clock.Never;
    private double _meowStart = -1;
    private const double MeowDuration = 1.4;
    private const double ShowSeconds = 8;

    /// `--demo-timers` cannot wait for a wall-clock minute to roll around.
    public void ArmDemoReminder(double seconds)
    {
        _demoFireAt = Clock.Now + seconds;
        if (_bus.Settings.ReminderText.Length == 0)
        {
            _bus.Settings.ReminderText = "Stand up and look out of a window.";
        }
    }

    public ModuleOutput Update(in TickContext ctx)
    {
        double now = Clock.Now;

        if (now >= _demoFireAt)
        {
            _demoFireAt = Clock.Never;
            Fire(_bus.Settings.ReminderText, "demo");
        }
        else if (_bus.Settings.ReminderEnabled && !_bus.Busy)
        {
            CheckClock(ctx.SecondsSinceKey);
        }

        if (_meowStart < 0) return ModuleOutput.None;
        double p = (now - _meowStart) / MeowDuration;
        if (p >= 1) { _meowStart = -1; return ModuleOutput.None; }

        // The meow: a head-forward lunge and a wobble. There is no open-mouth cell in
        // the rig and inventing one would need art in three themes, so the gesture
        // carries it.
        var outv = new ModuleOutput();
        double t = p;
        double envelope = (1 - t) * (1 - t);
        outv.Squash = 1 + envelope * 0.07 * Math.Sin(Math.PI * 4 * t);
        outv.Offset.Y = -envelope * _bus.Atlas.Wellness.BobHeight * 0.8
                        * Math.Sin(Math.PI * 2 * t);
        return outv;
    }

    private void CheckClock(double secondsSinceKey)
    {
        if (Parse(_bus.Settings.ReminderTime) is not { } target) return;

        // The one place the app asks for wall-clock time rather than the monotonic
        // Clock: "half past nine" is a question about the user's calendar, not about
        // how long the process has been up.
        var date = DateTime.Now;
        int dayKey = date.Year * 1000 + date.DayOfYear;
        if (date.Hour != target.H || date.Minute != target.M || _firedOnDay == dayKey) return;
        _firedOnDay = dayKey;

        if (secondsSinceKey > _bus.AwaySeconds)
        {
            // Marked as fired, then dropped: a reminder is about a moment, and
            // replaying it an hour later is worse than not having it.
            _bus.LogDemo($"reminder SKIPPED, away {secondsSinceKey:0}s");
            return;
        }
        Fire(_bus.Settings.ReminderText, $"clock {_bus.Settings.ReminderTime}");
    }

    private void Fire(string text, string why)
    {
        string message = text.Length == 0 ? "Reminder!" : text;
        _bus.Bubble?.Say(message, ShowSeconds);
        _bus.Chime();
        _meowStart = Clock.Now;
        _bus.LogDemo($"reminder ({why}) \"{message}\"");
    }

    // MARK: - pinned note

    public void Pin(string? text)
    {
        _bus.Bubble?.Pin(text);
        _bus.LogDemo(text is null ? "note cleared" : $"note pinned: \"{text}\"");
    }

    public void ClearReminder()
    {
        _bus.Settings.ReminderEnabled = false;
        _bus.Settings.ReminderTime = "";
        // Without this, re-enabling a reminder for a time that has already passed today
        // would be blocked by the day guard until midnight.
        _firedOnDay = -1;
    }

    // MARK: - parsing

    /// Public, not private: the settings window validates its time field with the same
    /// parser the tick fires on, so the two cannot come to disagree about what counts
    /// as a time.
    public static (int H, int M)? Parse(string s)
    {
        string[] parts = s.Trim().Split(':');
        if (parts.Length != 2) return null;
        if (!int.TryParse(parts[0], NumberStyles.None, CultureInfo.InvariantCulture, out int h))
            return null;
        if (!int.TryParse(parts[1], NumberStyles.None, CultureInfo.InvariantCulture, out int m))
            return null;
        if (h is < 0 or > 23 || m is < 0 or > 59) return null;
        return (h, m);
    }

    public static string Normalise(string s) =>
        Parse(s) is { } p ? $"{p.H:00}:{p.M:00}" : s;

    public static string DefaultTimeString() => DateTime.Now.ToString("HH:mm",
        CultureInfo.InvariantCulture);
}
